File: nexus/web/routes/memories.py
"""
Memory Routes - REST API endpoints for memory CRUD operations

Provides comprehensive memory management endpoints:
- List/search memories with filtering
- Create new memories
- Get specific memory by ID
- Update existing memories
- Delete memories
- Semantic search using embeddings
"""

import asyncio
from datetime import datetime, UTC
from typing import Dict, List, Any, Optional

from fastapi import APIRouter, HTTPException, Query, Path, status
from fastapi.responses import JSONResponse
from pydantic import BaseModel, Field
from loguru import logger

from ...server.nexus_manager import get_memory_manager
from ...database.managers import MemoryManager
from ..app import (
    MemoryCreateRequest,
    MemoryUpdateRequest,
    MemoryResponse,
    MemoryListResponse,
    MemoryCreateResponse,
    SemanticSearchRequest,
    SemanticSearchResponse,
)
from ..websocket.manager import broadcast_event


router = APIRouter()


# =============================================================================
# Memory CRUD Endpoints
# =============================================================================


@router.get(
    "/memories",
    response_model=MemoryListResponse,
    summary="List memories",
    description="Retrieve memories with optional filtering by query, category, and memory lane type"
)
async def list_memories(
    agent_type: str = Query(
        "general",
        description="Agent type to retrieve memories from",
        pattern="^(general|claude-code|gemini|qwen|amp|droid|opencode|codex)$"
    ),
    query: Optional[str] = Query(
        None,
        description="Search query for text filtering"
    ),
    category: Optional[str] = Query(
        None,
        description="Filter by category"
    ),
    memory_lane_type: Optional[str] = Query(
        None,
        description="Filter by Memory Lane type"
    ),
    limit: int = Query(
        20,
        ge=1,
        le=100,
        description="Maximum number of results"
    ),
    offset: int = Query(
        0,
        ge=0,
        description="Offset for pagination"
    )
):
    """
    List memories with optional filtering.

    Supports filtering by:
    - agent_type: The agent namespace
    - query: Text search in memory content
    - category: Memory category filter
    - memory_lane_type: Memory Lane cognitive type filter
    """
    try:
        manager = get_memory_manager()
        await manager.ensure_initialized()

        # Use text-based search if query provided
        if query:
            result = await manager.memory_manager.search_memories(
                query=query,
                agent_type=agent_type,
                limit=limit,
                category=category,
                memory_lane_type=memory_lane_type
            )
        else:
            # Direct listing without text search
            result = await manager.memory_manager.search_memories(
                query="",  # Empty query returns all
                agent_type=agent_type,
                limit=limit,
                category=category,
                memory_lane_type=memory_lane_type
            )

        if not result.get("success"):
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=result.get("error", "Unknown error")
            )

        return MemoryListResponse(
            success=True,
            total=result.get("total", 0),
            results=[
                MemoryResponse(
                    id=m["id"],
                    content=m["content"],
                    category=m["category"],
                    category_description=m.get("category_description"),
                    memory_lane_type=m.get("memory_lane_type"),
                    labels=m.get("labels", []),
                    metadata=m.get("metadata", {}),
                    similarity_score=m.get("similarity_score"),
                    relevance_score=m.get("relevance_score"),
                    created_at=m["created_at"],
                    last_accessed=m.get("last_accessed"),
                    access_count=m["access_count"]
                )
                for m in result.get("results", [])
            ],
            query=query,
            agent_type=agent_type,
            filters={
                "category": category,
                "memory_lane_type": memory_lane_type,
            }
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error listing memories: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to list memories: {str(e)}"
        )


@router.post(
    "/memories",
    response_model=MemoryCreateResponse,
    summary="Create a memory",
    description="Store a new memory with automatic categorization and embedding generation",
    status_code=status.HTTP_201_CREATED
)
async def create_memory(request: MemoryCreateRequest):
    """
    Create a new memory.

    The memory will be:
    - Validated for content and category
    - Stored in the database
    - Optionally embedded for semantic search (if enabled)
    - Broadcast to WebSocket subscribers
    """
    try:
        manager = get_memory_manager()
        await manager.ensure_initialized()

        result = await manager.store_memory(
            content=request.content,
            agent_type=request.agent_type,
            category=request.category,
            labels=request.labels,
            metadata={
                **request.metadata,
                "memory_lane_type": request.memory_lane_type
            }
        )

        if not result.get("success"):
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=result.get("error", "Failed to create memory")
            )

        # Broadcast memory created event
        await broadcast_event({
            "type": "memory_created",
            "data": {
                "memory_id": result.get("memory_id"),
                "agent_type": request.agent_type,
                "category": request.category,
                "content": request.content[:200] + "..." if len(request.content) > 200 else request.content,
                "created_at": datetime.now(UTC).isoformat()
            }
        })

        return MemoryCreateResponse(
            success=True,
            memory_id=result.get("memory_id"),
            agent_type=request.agent_type,
            category=request.category
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating memory: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create memory: {str(e)}"
        )


@router.get(
    "/memories/{memory_id}",
    response_model=MemoryResponse,
    summary="Get a specific memory",
    description="Retrieve detailed information about a specific memory by ID"
)
async def get_memory(
    memory_id: int = Path(
        ...,
        ge=1,
        description="Memory ID"
    ),
    agent_type: str = Query(
        "general",
        description="Agent type for the memory"
    )
):
    """Get a specific memory by ID"""
    try:
        manager = get_memory_manager()
        await manager.ensure_initialized()

        # Search specifically for this memory ID
        result = await manager.memory_manager.search_memories(
            query="",
            agent_type=agent_type,
            limit=1
        )

        if not result.get("success"):
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail=result.get("error")
            )

        # Find the specific memory
        memory_data = None
        for m in result.get("results", []):
            if m["id"] == memory_id:
                memory_data = m
                break

        if not memory_data:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"Memory {memory_id} not found"
            )

        return MemoryResponse(
            id=memory_data["id"],
            content=memory_data["content"],
            category=memory_data["category"],
            category_description=memory_data.get("category_description"),
            memory_lane_type=memory_data.get("memory_lane_type"),
            labels=memory_data.get("labels", []),
            metadata=memory_data.get("metadata", {}),
            similarity_score=memory_data.get("similarity_score"),
            relevance_score=memory_data.get("relevance_score"),
            created_at=memory_data["created_at"],
            last_accessed=memory_data.get("last_accessed"),
            access_count=memory_data["access_count"]
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error getting memory: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get memory: {str(e)}"
        )


@router.put(
    "/memories/{memory_id}",
    response_model=MemoryResponse,
    summary="Update a memory",
    description="Update an existing memory's content, labels, or metadata"
)
async def update_memory(
    memory_id: int = Path(..., ge=1),
    request: MemoryUpdateRequest = None
):
    """Update a memory"""
    try:
        manager = get_memory_manager()
        await manager.ensure_initialized()

        # Get the memory first
        db_manager = manager.db_manager
        async with db_manager.get_async_session() as session:
            from ...database.models import Memory
            from sqlalchemy import select

            stmt = select(Memory).where(Memory.id == memory_id)
            result = await session.execute(stmt)
            memory = result.scalar_one_or_none()

            if not memory:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Memory {memory_id} not found"
                )

            # Update fields
            if request.content is not None:
                memory.content = request.content
            if request.category is not None:
                memory.category = request.category
            if request.labels is not None:
                memory.labels = request.labels
            if request.metadata is not None:
                memory.extra_metadata = request.metadata
            if request.memory_lane_type is not None:
                memory.memory_lane_type = request.memory_lane_type
            if request.is_active is not None:
                memory.is_active = request.is_active
            if request.is_archived is not None:
                memory.is_archived = request.is_archived

            memory.updated_at = datetime.now(UTC)
            await session.commit()
            await session.refresh(memory)

        # Broadcast memory updated event
        await broadcast_event({
            "type": "memory_updated",
            "data": {
                "memory_id": memory_id,
                "updated_at": datetime.now(UTC).isoformat()
            }
        })

        return MemoryResponse(
            id=memory.id,
            content=memory.content,
            category=memory.category,
            category_description=None,
            memory_lane_type=memory.memory_lane_type,
            labels=memory.labels or [],
            metadata=memory.extra_metadata or {},
            similarity_score=memory.similarity_score,
            relevance_score=memory.relevance_score,
            created_at=memory.created_at.isoformat(),
            last_accessed=memory.last_accessed.isoformat() if memory.last_accessed else None,
            access_count=memory.access_count
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error updating memory: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update memory: {str(e)}"
        )


@router.delete(
    "/memories/{memory_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Delete a memory",
    description="Soft-delete a memory (marks as inactive)"
)
async def delete_memory(
    memory_id: int = Path(..., ge=1)
):
    """Delete a memory (soft delete)"""
    try:
        manager = get_memory_manager()
        await manager.ensure_initialized()

        db_manager = manager.db_manager
        async with db_manager.get_async_session() as session:
            from ...database.models import Memory
            from sqlalchemy import select, update

            stmt = select(Memory).where(Memory.id == memory_id)
            result = await session.execute(stmt)
            memory = result.scalar_one_or_none()

            if not memory:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Memory {memory_id} not found"
                )

            # Soft delete
            memory.is_active = False
            memory.is_archived = True
            memory.updated_at = datetime.now(UTC)
            await session.commit()

        # Broadcast memory deleted event
        await broadcast_event({
            "type": "memory_deleted",
            "data": {
                "memory_id": memory_id,
                "deleted_at": datetime.now(UTC).isoformat()
            }
        })

        return JSONResponse(
            status_code=status.HTTP_204_NO_CONTENT,
            content=None
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting memory: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete memory: {str(e)}"
        )


# =============================================================================
# Semantic Search Endpoints
# =============================================================================


@router.post(
    "/search/semantic",
    response_model=SemanticSearchResponse,
    summary="Semantic search",
    description="Search memories using vector embeddings for semantic similarity"
)
async def semantic_search(request: SemanticSearchRequest):
    """
    Perform semantic search using vector embeddings.

    Requires sqlite-vec to be installed and embeddings enabled.
    Returns memories ranked by semantic similarity to the query.
    """
    try:
        manager = get_memory_manager()
        await manager.ensure_initialized()

        result = await manager.memory_manager.search_memories_by_embedding(
            query=request.query,
            agent_type=request.agent_type,
            k=request.k,
            threshold=request.threshold,
            category=request.category,
            memory_lane_type=request.memory_lane_type
        )

        if not result.get("success"):
            return SemanticSearchResponse(
                success=False,
                results=[],
                total=0,
                query=request.query,
                agent_type=request.agent_type,
                filters={
                    "category": request.category,
                    "memory_lane_type": request.memory_lane_type,
                    "threshold": request.threshold,
                },
                error=result.get("error", "Search failed")
            )

        return SemanticSearchResponse(
            success=True,
            results=result.get("results", []),
            total=result.get("total", 0),
            query=request.query,
            agent_type=request.agent_type,
            filters={
                "category": request.category,
                "memory_lane_type": request.memory_lane_type,
                "threshold": request.threshold,
            }
        )

    except Exception as e:
        logger.error(f"Error in semantic search: {e}")
        return SemanticSearchResponse(
            success=False,
            results=[],
            total=0,
            query=request.query,
            agent_type=request.agent_type,
            filters={
                "category": request.category,
                "memory_lane_type": request.memory_lane_type,
                "threshold": request.threshold,
            },
            error=str(e)
        )

File: crates/nexus-web/src/api/memories.rs
//! Memory CRUD API endpoints

use axum::{
    extract::{Path, Query, State},
    http::StatusCode,
    Json,
};
use chrono::Utc;
use nexus_storage::repository::StoreMemoryParams;
use serde::Deserialize;
use serde_json::json;
use std::sync::Arc;
use tokio::sync::RwLock;
use tracing::{info, warn};

use crate::{
    error::{Result, WebError},
    models::{
        CreateMemoryRequest, MemoryCreateResponse, MemoryListResponse, MemoryResponse,
        SearchRequest, SearchResponse, UpdateMemoryRequest, WebSocketMessage,
    },
    state::AppState,
};

/// Query parameters for listing memories
#[derive(Debug, Deserialize)]
pub struct ListMemoriesQuery {
    #[serde(default = "default_agent_type")]
    pub agent_type: String,
    pub query: Option<String>,
    pub category: Option<String>,
    pub memory_lane_type: Option<String>,
    #[serde(default = "default_limit")]
    pub limit: usize,
    #[serde(default)]
    pub offset: usize,
}

fn default_agent_type() -> String {
    "general".to_string()
}

fn default_limit() -> usize {
    20
}

/// List memories with optional filtering
pub async fn list_memories(
    State(state): State<Arc<RwLock<AppState>>>,
    Query(params): Query<ListMemoriesQuery>,
) -> Result<Json<MemoryListResponse>> {
    let state = state.read().await;

    // Get or create namespace
    let namespace = state
        .namespace_repo
        .get_or_create(&params.agent_type, &params.agent_type)
        .await?;

    // Search memories
    let memories = state
        .memory_repo
        .search_by_namespace(namespace.id, params.limit, params.offset)
        .await?;

    let total = state.memory_repo.count_by_namespace(namespace.id).await?;

    // Convert to response models
    let results: Vec<MemoryResponse> = memories.into_iter().map(MemoryResponse::from).collect();

    let filters = json!({
        "category": params.category,
        "memory_lane_type": params.memory_lane_type,
    });

    Ok(Json(MemoryListResponse {
        success: true,
        total,
        results,
        query: params.query,
        agent_type: params.agent_type,
        filters,
    }))
}

/// Create a new memory
pub async fn create_memory(
    State(state): State<Arc<RwLock<AppState>>>,
    Json(request): Json<CreateMemoryRequest>,
) -> Result<(StatusCode, Json<MemoryCreateResponse>)> {
    let state = state.read().await;

    // Validate content
    if request.content.trim().is_empty() {
        return Err(WebError::InvalidRequest(
            "Content cannot be empty".to_string(),
        ));
    }

    // Get or create namespace
    let namespace = state
        .namespace_repo
        .get_or_create(&request.agent_type, &request.agent_type)
        .await?;

    // Store memory
    let memory = state
        .memory_repo
        .store(StoreMemoryParams {
            namespace_id: namespace.id,
            content: &request.content,
            category: &request.category,
            memory_lane_type: request.memory_lane_type.as_ref(),
            labels: &request.labels,
            metadata: &request.metadata,
            embedding: None,
            embedding_model: None,
        })
        .await?;

    // Broadcast to WebSocket clients
    let memory_response = MemoryResponse::from(memory.clone());
    let ws_msg = WebSocketMessage::memory_stored(&memory_response, &request.agent_type);
    let _ = state.broadcast_ws(ws_msg);

    info!(
        "Memory created: id={}, agent_type={}",
        memory.id, request.agent_type
    );

    Ok((
        StatusCode::CREATED,
        Json(MemoryCreateResponse {
            success: true,
            memory_id: Some(memory.id),
            agent_type: request.agent_type,
            category: request.category.to_string(),
            error: None,
        }),
    ))
}

/// Get a specific memory by ID
pub async fn get_memory(
    State(state): State<Arc<RwLock<AppState>>>,
    Path(id): Path<i64>,
) -> Result<Json<MemoryResponse>> {
    let state = state.read().await;

    let memory = state
        .memory_repo
        .get_by_id(id)
        .await?
        .ok_or_else(|| WebError::NotFound(format!("Memory {} not found", id)))?;

    // Update access count
    let _ = state.memory_repo.touch(id).await;

    Ok(Json(MemoryResponse::from(memory)))
}

/// Update an existing memory
pub async fn update_memory(
    State(state): State<Arc<RwLock<AppState>>>,
    Path(id): Path<i64>,
    Json(request): Json<UpdateMemoryRequest>,
) -> Result<Json<MemoryResponse>> {
    let state = state.read().await;

    // Check if memory exists
    let existing = state
        .memory_repo
        .get_by_id(id)
        .await?
        .ok_or_else(|| WebError::NotFound(format!("Memory {} not found", id)))?;

    // Build update query dynamically
    let mut updates: Vec<String> = Vec::new();

    if let Some(content) = request.content {
        if !content.trim().is_empty() {
            updates.push(format!("content = '{}'", content.replace("'", "''")));
        }
    }

    if let Some(category) = request.category {
        updates.push(format!("category = '{}'", category));
    }

    if let Some(memory_lane_type) = request.memory_lane_type {
        updates.push(format!("memory_lane_type = '{}'", memory_lane_type));
    }

    if let Some(labels) = request.labels {
        match serde_json::to_string(&labels) {
            Ok(labels_json) => {
                updates.push(format!("labels = '{}'", labels_json.replace("'", "''")));
            }
            Err(e) => {
                warn!(error = %e, "Failed to serialize labels for memory update; labels omitted from SQL update");
                // Do NOT push an empty-string assignment — that would corrupt the row.
            }
        }
    }

    if let Some(metadata) = request.metadata {
        match serde_json::to_string(&metadata) {
            Ok(metadata_json) => {
                updates.push(format!("metadata = '{}'", metadata_json.replace("'", "''")));
            }
            Err(e) => {
                warn!(error = %e, "Failed to serialize metadata for memory update; metadata omitted from SQL update");
                // Do NOT push an empty-string assignment — that would corrupt the row.
            }
        }
    }

    if let Some(is_active) = request.is_active {
        updates.push(format!("is_active = {}", if is_active { 1 } else { 0 }));
    }

    if let Some(is_archived) = request.is_archived {
        updates.push(format!("is_archived = {}", if is_archived { 1 } else { 0 }));
    }

    if updates.is_empty() {
        return Ok(Json(MemoryResponse::from(existing)));
    }

    updates.push(format!("updated_at = '{}'", Utc::now().to_rfc3339()));

    let query = format!("UPDATE memories SET {} WHERE id = ?", updates.join(", "));

    sqlx::query(&query)
        .bind(id)
        .execute(state.pool())
        .await
        .map_err(|e| WebError::Storage(e.to_string()))?;

    // Fetch updated memory
    let updated = state
        .memory_repo
        .get_by_id(id)
        .await?
        .ok_or_else(|| WebError::NotFound(format!("Memory {} not found after update", id)))?;

    // Broadcast update
    let ws_msg = WebSocketMessage::memory_updated(id);
    let _ = state.broadcast_ws(ws_msg);

    info!("Memory updated: id={}", id);

    Ok(Json(MemoryResponse::from(updated)))
}

/// Delete a memory (soft delete)
pub async fn delete_memory(
    State(state): State<Arc<RwLock<AppState>>>,
    Path(id): Path<i64>,
) -> Result<StatusCode> {
    let state = state.read().await;

    // Check if memory exists
    let _ = state
        .memory_repo
        .get_by_id(id)
        .await?
        .ok_or_else(|| WebError::NotFound(format!("Memory {} not found", id)))?;

    // Soft delete: mark as inactive and archived
    sqlx::query("UPDATE memories SET is_active = 0, is_archived = 1, updated_at = ? WHERE id = ?")
        .bind(Utc::now())
        .bind(id)
        .execute(state.pool())
        .await
        .map_err(|e| WebError::Storage(e.to_string()))?;

    // Broadcast deletion
    let ws_msg = WebSocketMessage::memory_deleted(id);
    let _ = state.broadcast_ws(ws_msg);

    info!("Memory deleted: id={}", id);

    Ok(StatusCode::NO_CONTENT)
}

/// Search memories using semantic or text search
pub async fn search_memories(
    State(state): State<Arc<RwLock<AppState>>>,
    Json(request): Json<SearchRequest>,
) -> Result<Json<SearchResponse>> {
    let state = state.read().await;

    // Validate query
    if request.query.trim().is_empty() {
        return Err(WebError::InvalidRequest(
            "Query cannot be empty".to_string(),
        ));
    }

    // Get namespace
    let namespace = state
        .namespace_repo
        .get_or_create(&request.agent_type, &request.agent_type)
        .await?;

    // For now, use text-based search (semantic search would require embeddings)
    // Search in content using LIKE
    let search_pattern = format!(
        "%{}%",
        request.query.replace("%", "\\%").replace("_", "\\_")
    );

    let query_str = "SELECT * FROM memories WHERE namespace_id = ? AND is_active = 1 AND content LIKE ? ORDER BY created_at DESC LIMIT ? OFFSET ?".to_string();

    let rows: Vec<nexus_storage::models::MemoryRow> = sqlx::query_as(&query_str)
        .bind(namespace.id)
        .bind(&search_pattern)
        .bind(request.limit as i64)
        .bind(request.offset as i64)
        .fetch_all(state.pool())
        .await
        .map_err(|e| WebError::Storage(e.to_string()))?;

    // Convert rows to memories
    let memories: Vec<nexus_core::Memory> = rows
        .into_iter()
        .map(row_to_memory)
        .collect::<crate::error::Result<Vec<_>>>()?;

    let results: Vec<MemoryResponse> = memories.into_iter().map(MemoryResponse::from).collect();

    let total = results.len() as i64;

    let filters = json!({
        "category": request.category.map(|c| c.to_string()),
        "memory_lane_type": request.memory_lane_type.map(|t| t.to_string()),
        "threshold": request.threshold,
    });

    Ok(Json(SearchResponse {
        success: true,
        results,
        total,
        query: request.query,
        agent_type: request.agent_type,
        filters,
        error: None,
    }))
}

/// Convert a database row to a Memory
fn row_to_memory(
    row: nexus_storage::models::MemoryRow,
) -> crate::error::Result<nexus_core::Memory> {
    use nexus_core::{Memory, MemoryCategory, MemoryLaneType};

    let labels: Vec<String> = serde_json::from_str(&row.labels).map_err(|e| {
        crate::error::WebError::Storage(format!("corrupted labels JSON for memory {}: {e}", row.id))
    })?;
    let metadata: serde_json::Value = serde_json::from_str(&row.metadata).map_err(|e| {
        crate::error::WebError::Storage(format!(
            "corrupted metadata JSON for memory {}: {e}",
            row.id
        ))
    })?;
    let embedding: Option<Vec<f32>> = row
        .content_embedding
        .map(|e| {
            serde_json::from_str(&e).map_err(|err| {
                crate::error::WebError::Storage(format!(
                    "corrupted embedding JSON for memory {}: {err}",
                    row.id
                ))
            })
        })
        .transpose()?;

    Ok(Memory {
        id: row.id,
        namespace_id: row.namespace_id,
        content: row.content,
        category: MemoryCategory::parse(&row.category).ok_or_else(|| {
            WebError::Storage(format!(
                "Unknown memory category '{}' persisted in database; row may be corrupted",
                row.category
            ))
        })?,
        memory_lane_type: match &row.memory_lane_type {
            Some(s) => Some(MemoryLaneType::parse(s).ok_or_else(|| {
                WebError::Storage(format!(
                    "Unknown memory_lane_type '{}' persisted in database; row may be corrupted",
                    s
                ))
            })?),
            None => None,
        },
        labels,
        metadata,
        similarity_score: row.similarity_score,
        relevance_score: row.relevance_score,
        content_embedding: embedding,
        embedding_model: row.embedding_model,
        created_at: row.created_at,
        updated_at: row.updated_at,
        last_accessed: row.last_accessed,
        is_active: row.is_active,
        is_archived: row.is_archived,
        access_count: row.access_count,
    })
}
